src/codegen.cpp

```diff
@@ -1024,8 +1031,8 @@ static const auto jl_alloc_obj_func = new JuliaFunction<TypeFnContextAndSizeT>{
 #endif
         FnAttrs.addAttribute(Attribute::WillReturn);
         FnAttrs.addAttribute(Attribute::NoUnwind);
-        FnAttrs.addAttribute(Attribute::MustProgress);
-        // FnAttrs.addAttribute(Attribute::InaccessibleMemOnly); // Maybe readnone/readonly instead?
+        FnAttrs.addAttribute(Attribute::NoFree);
```

Comment:
This function does free memory and is therefore probably also a sync point

Reply:
nofree means 
> nofree
This indicates that callee does not free the pointer argument. This is not a valid attribute for return values.
which is true, and thinking about it. Every function call that potentially has a safepoint syncs so nosync isn't very legal.

Reply:
> This function attribute indicates that the function does not, directly or transitively, call a memory-deallocation function (free, for example) on a memory allocation which existed before the call.

> As a result, uncaptured pointers that are known to be dereferenceable prior to a call to a function with the nofree attribute are still known to be dereferenceable after the call. The capturing condition is necessary in environments where the function might communicate the pointer to another thread which then deallocates the memory. Alternatively, nosync would ensure such communication cannot happen and even captured pointers cannot be freed by the function.

> A nofree function is explicitly allowed to free memory which it allocated or (if not nosync) arrange for another thread to free memory on it’s behalf. As a result, perhaps surprisingly, a nofree function can return a pointer to a previously deallocated memory object.

So I guess I'm wondering what it means to be known as dereferenceable. If a GC pointer was legal before this call it should still be valid after it. Then again I'm not really sure if this allows for too many more optimizations.